### ISSUE TITLE
Experiment using concurrent garbage collector

### ISSUE THREAD

**Issue by mpickering** [haskell/haskell-language-server#789]
If a GC happens during a request there can be a reasonably big pause if you have a largish heap. 

We should try using the concurrent GC to attempt to reduce pause times.

**Comment by pepeiborra** [comment]
I have checked again with ghc 8.10.3 and it seems to work pretty well. The crashes are gone. I have seen one crash with `--nonmoving-gc -A128M` enabled, but it's extremely hard to reproduce and could be something else:

```
ghcide: internal error: SMALL_MUT_ARR_PTRS_FROZEN_CLEAN object (0x4228156f38) entered!
    (GHC version 8.10.3 for x86_64_apple_darwin)
    Please report this as a GHC bug:  https://www.haskell.org/ghc/reportabug
```

I also collected some performance numbers using the benchmark suite. Overall, the timings were similar or slightly worse in most benchmarks. I wasn't able to find a set of GC flags using `--nonmoving-gc` that showed an improvement over the ones we use right now (`-A128M -qg -I0`). But I did notice that `-qg`, which disables the parallel GC, is a net loss - all the benchmarks are faster without it.   

For the edit benchmark, which performs an edit and then waits for diagnostics, over 100 samples, the chart below shows the live bytes over time (as reported by `-S`) for various configurations

- upstream: `-A128M -qg -I0` 
- Adefault: `-qg -I0`
- parallelGC: `-A128M -I0`
- A64:  `-A64M -qg -I0`
- nmA64: `--nonmoving-gc -A64M`
- nmAdefault: `-qg -I0 --nonmoving-gc`

![image](https://user-images.githubusercontent.com/26626/103491354-177bb980-4e1b-11eb-8a02-f217650044a0.png)

Branch to reproduce: https://github.com/pepeiborra/ide/tree/benchmark-rts-opts-nm

**Comment by mpickering** [comment]
@pepeiborra Thanks for looking into this. I will ask about the panic you are seeing. 

Isn't the idea behind using the nonmoving-gc to reduce the pause times? This is interesting for us because if the pause happens when serving a request then the user will notice it. For example, if you hover, then a GC kicks in for 1-2s then the hover response will also be delayed. With the nonmoving-gc then the pause will be shorter and therefore a smoother experience for the user, even if it's slightly slower.

**Comment by pepeiborra** [comment]
Yes, that's correct. To measure pauses the benchmark suite needs to be extended to show not max time in addition to total time.

**Comment by fishtreesugar** [comment]
https://twitter.com/monadiccheng/status/1539583255317446658
> haskell-language-server with --nonmoving-gc is a lot smoother, when heap size goes beyond 10GiB

**Comment by hasufell** [comment]
Since this works, can someone raise a PR to add this to the documentation for experimentation friendly users?

**Comment by TeofilC** [comment]
Same as https://github.com/haskell/haskell-language-server/issues/3656

**Comment by crtschin** [comment]
Closing. More active discussion has occurred in duplicate issue https://github.com/haskell/haskell-language-server/issues/3656.
